tests/test_evalscript.py

```diff
@@ -0,0 +1,136 @@
+    merged_output = "merged_bands"
+    evalscript = generate_evalscript(
+        data_collection=data_collection, bands=bands, meta_bands=meta_bands, merged_output=merged_output
+    )
+
+    expected_bands_output_spec = "{ " + f'id: "{merged_output}", bands: {len(bands)}'
+    assert expected_bands_output_spec in evalscript
+
+    expected_bands_return_spec = f"{merged_output}: [" + ", ".join(f"sample.{b}" for b in bands) + "]"
+    assert expected_bands_return_spec in evalscript
+
+
+def test_units() -> None:
+    bands = ["B04", "B03", "B02", "B02"]
+    data_collection = DataCollection.SENTINEL2_L1C
+
+    evalscript_dn = generate_evalscript(data_collection=data_collection, bands=bands)
+    assert evalscript_dn.count('"DN"') == len(bands)
+    assert evalscript_dn.count('"REFLECTANCE"') == 0
+
+    evalscript_reflectance = generate_evalscript(data_collection=data_collection, bands=bands, use_dn=False)
+    assert evalscript_reflectance.count('"REFLECTANCE"') == len(bands)
+    assert evalscript_reflectance.count('"DN"') == 0
+
+
+@pytest.mark.parametrize("use_dn", [True, False], ids=lambda x: f"use_dn: {x}")
+def test_sample_type(use_dn: bool) -> None:
+    data_collection = DataCollection.SENTINEL2_L1C
+    evalscript = generate_evalscript(data_collection=data_collection, use_dn=use_dn)
+
+    expected_uint_count = len(data_collection.bands) if use_dn else 0
+    assert evalscript.count('"UINT16"') == expected_uint_count
+
+    expected_float_count = len(data_collection.bands) if not use_dn else 0
+    assert evalscript.count('"FLOAT32"') == expected_float_count
+
+
+@pytest.mark.parametrize("use_dn", [True, False], ids=lambda x: f"use_dn: {x}")
+def test_sample_type_merged(use_dn: bool) -> None:
+    evalscript = generate_evalscript(data_collection=DataCollection.SENTINEL2_L1C, merged_output="bands", use_dn=use_dn)
+
+    expected_uint_count = 1 if use_dn else 0
+    assert evalscript.count('"UINT16"') == expected_uint_count
+
+    expected_float_count = 1 if not use_dn else 0
+    assert evalscript.count('"FLOAT32"') == expected_float_count
+
+
+@pytest.mark.parametrize("data_collection", [DataCollection.LANDSAT_TM_L2, DataCollection.SENTINEL2_L2A])
+@pytest.mark.parametrize("merged_output", [None, "bands"])
+@pytest.mark.parametrize("use_dn", [True, False])
+def test_valid_evalscript(data_collection: DataCollection, merged_output: Optional[str], use_dn: bool) -> None:
+    bands = ["B05", "B04", "B03"]
+    meta_bands = ["dataMask"]
+    evalscript = generate_evalscript(
+        data_collection=data_collection,
+        bands=bands,
+        meta_bands=meta_bands,
+        merged_output=merged_output,
+        use_dn=use_dn,
+    )
+    if merged_output is not None:
+        responses = [SentinelHubRequest.output_response(merged_output, MimeType.TIFF)]
+        responses.extend([SentinelHubRequest.output_response(band, MimeType.TIFF) for band in meta_bands])
```

Comment:
i see, this merges together bands, but not meta bands as well?

Reply:
indeed. meta bands are treated differently already in the data collection, so I thought it wouldn't make sense to treat them the same as bands here. I imagine it's more common to want to merge together bands separately. 

Additionally, this could then be simplified in eo-learn, where we request metabands separately, but ultimately manually pack the bands together into a single image. This code would now become redundant, because the band packing is done on the server side already.

Reply:
So in `eo-learn` we download each band separately? What are the benefits of merging them together on the server side? does it cost less / takes up less space?

Reply:
A question for core, or perhaps @batic can help out, but what I do know is that it's common to pack the bands into a single output. One additional limitation is the JPEG/PNG only work with 1 or 3 bands, but here we use TIFF mimetype anyway, so it doesn't really make a difference from that point

Reply:
From the point of the service, no difference. Each band separately (e.g., `response["B01.tif"]`, `response["B02.tif"]` ...), or all the bands in one output (e.g. x-bands in `response["default.tif"]`) is the same. The only problem might come if one would pack tens (or hundreds) of bands into a single tiff. The PUs are computed based on the number of inputs. 

IIRC, we download bands separately in eo-learn to reduce the download size, as we could request `DN`s for most of the bands, and `bool` for masks, `uint`s for some other data etc, and later convert them to `float` and pack them into a single feature. It probably still doesn't make sense, as typically you'd join together bands of the same type (e.g. Sentinel-2 bands into one feature, masks into masks, angles etc into its own feature etc).

Reply:
> IIRC, we download bands separately in eo-learn to reduce the download size, as we could request DNs for most of the bands, and bool for masks, uints for some other data etc, and later convert them to float and pack them into a single feature. It probably still doesn't make sense, as typically you'd join together bands of the same type (e.g. Sentinel-2 bands into one feature, masks into masks, angles etc into its own feature etc).

With the changes proposed here, you are now able to download all bands as DN, already packed together into a single output, while still having other features (but only meta bands such as CLM, CLM, dataMask, ..) as separate response objects, so all good from this point of view as well.

thanks!

Reply:
Looking at [`S2-L1C`](https://docs.sentinel-hub.com/api/latest/data/sentinel-2-l1c/#units) band units, the angles (which are eo-learn `data` features), are not in DNs. Is this taken into account?

Reply:
Yes, the `use_dn` parameter is more of a `use_dn_if_possible_otherwise_the_first_choice`, which also covers the case of `dont_use_dn_unless_they_are_the_only_possible_choice`

accepting better name suggestions :D

Reply:
`prefer_dn`?

Reply:
Yes, most likely something along those lines. I was also thinking about `prioritize_dn`, but it's longer

Reply:
that one is good, i'd go with that. it's more formal :)

but i'm stil on the verge for merging bands. If its of no benefit to us, i'm not sure why we'd try and do it in the evalscript :thinking: it makes the evalscript function quite a bit more complex, so i'd want a trade-off

Reply:
> but i'm stil on the verge for merging bands. If its of no benefit to us, i'm not sure why we'd try and do it in the evalscript 🤔 it makes the evalscript function quite a bit more complex, so i'd want a trade-off

well, there's no benefit from the memory/speed/cost point of view, but there is a technical cost of manually packing the bands together. I think the code for supporting this in the evalscript is much nicer and less complex than the code needed to pack the actually data together manually:
- much of the code in [eo-learn](https://github.com/sentinel-hub/eo-learn/blob/develop/io/eolearn/io/sentinelhub_process.py#L593-L634) would become redundant due to this
- the [ExplodeBandsTask](https://github.com/sentinel-hub/eo-learn/blob/develop/core/eolearn/core/core_tasks.py#L517-L538) might become redundant as well at some point in the future, if we tweak the SHIT task in 4.0 so that it can also accept the option of providing bands separately

Reply:
> that one is good, i'd go with that. it's more formal :)

done

>  it makes the evalscript function quite a bit more complex, so i'd want a trade-off

I checked again and tried to make it a bit less complex/bloated and more intuitive in the last commit, please take a look when you manage

Reply:
i actually like it more before, i don't like changing state in the for loops (unless it's appending or something) :see_no_evil: 

what if you gather the information for output in a separate list/datastructure. You need a `band_name: sample_type` map right? Then after this loop you branch on whether the output is merged or not, and either you do a `.join` or just an f-string. Ping me on mattermost if i'm unclear.

Reply:
agreed. did what you suggested and it's much more elegant now. I need to get used to seeing more clearly what is the obstacle and trying to extract the information beforehand, so that the flow later on is easier to follow.

thanks!